qlib/contrib/model/pytorch_hgats_ts.py

```diff
@@ -0,0 +1,529 @@
+            d_feat=self.d_feat,
+            hidden_size=self.hidden_size,
+            num_layers=self.num_layers,
+            dropout=self.dropout,
+            base_model=self.base_model,
+        )
+        self.logger.info("model:\n{:}".format(self.GAT_model))
+        self.logger.info(
+            "model size: {:.4f} MB".format(count_parameters(self.GAT_model))
+        )
+
+        if optimizer.lower() == "adam":
+            self.train_optimizer = optim.Adam(self.GAT_model.parameters(), lr=self.lr)
+        elif optimizer.lower() == "gd":
+            self.train_optimizer = optim.SGD(self.GAT_model.parameters(), lr=self.lr)
+        else:
+            raise NotImplementedError(
+                "optimizer {} is not supported!".format(optimizer)
+            )
+
+        self.fitted = False
+        self.GAT_model.to(self.device)
+
+        self.bigG = pd.read_pickle("benchmarks/HGATs/hypergraph/CSI300.pkl")
+        self.num_ind = 29
+        # self.num_ind = (
+        #     pd.get_dummies(self.bigG.values.squeeze()).values.astype("float64").shape[1]
+        # )
+
+        # self.bigG = pd.read_pickle("benchmarks/GATs/Ind/bigG_trivial.pkl")
+        # `bigG_trivial` corresponds to the trivial situation where all stocks belong to differnet industries
+
+    @property
+    def use_gpu(self):
+        return self.device != torch.device("cpu")
+
+    def mse(self, pred, label):
+        loss = (pred - label) ** 2
+        return torch.mean(loss)
+
+    def loss_fn(self, pred, label):
+        mask = ~torch.isnan(label)
+
+        if self.loss == "mse":
+            return self.mse(pred[mask], label[mask])
+
+        raise ValueError("unknown loss `%s`" % self.loss)
+
+    def metric_fn(self, pred, label):
+
+        mask = torch.isfinite(label)
+
+        if self.metric == "" or self.metric == "loss":
+            return -self.loss_fn(pred[mask], label[mask])
+
+        raise ValueError("unknown metric `%s`" % self.metric)
+
+    def get_daily_inter(self, df, shuffle=False):
+        # organize the train data into daily batches
+        daily_count = df.groupby(level=0).size().values
+        daily_index = np.roll(np.cumsum(daily_count), 1)
+        daily_index[0] = 0
+        if shuffle:
+            # shuffle data
+            daily_shuffle = list(zip(daily_index, daily_count))
+            np.random.shuffle(daily_shuffle)
+            daily_index, daily_count = zip(*daily_shuffle)
+        return daily_index, daily_count
+
+    def train_epoch(self, data_loader, tsds, big_G):
+
+        self.GAT_model.train()
+
+        for data, i in data_loader:
+
+            data = data.squeeze()
+            feature = data[:, :, 0:-1].to(self.device)
+            label = data[:, -1, -1].to(self.device)
+
+            GH = pd.get_dummies(big_G.values.squeeze()).values.astype("float64")
+            GH = GH[
+                big_G.index.isin(
+                    tsds.get_index()[i.tolist()].to_frame()["instrument"].values
```

Comment:
Let's have a discussion of the content of big_G

Reply:
This is resolved - the industry information is now drawn from the daily stock batch.

Reply:
The problem with the mis-interpretation of x_j is also addressed.